File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Singapore.m
//
//  UIImage+Singapore.m
//  CountriesFlags
//

#import "Countries/UIImage+Singapore.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Singapore)

+ (UIImage *)flagSingaporeWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:238.0f/255.0f
                                        green:37.0f/255.0f
                                         blue:54.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height / 2.0f));
    
    [colorWhite setFill];
    CGFloat starRadius = size.height / 36.0f * 6.625f;
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.width / 54.0f * 11.46f, size.height / 36.0f * 9.0f)
                                    radius:starRadius
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorRed setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.width / 54.0f * 14.44f, size.height / 36.0f * 9.0f)
                                    radius:starRadius
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorWhite setFill];
    starRadius = size.height / 36.0f * 3.45f / 2.0f;
    // top
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 54.0f * 15.05f, size.height / 36.0f * 5.2f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // left middle
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 54.0f * 11.4f, size.height / 36.0f * 7.85f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // right middle
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 54.0f * 18.65f, size.height / 36.0f * 7.85f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // left bottom
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 54.0f * 12.8f, size.height / 36.0f * 12.1f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // right bottom
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 54.0f * 17.25f, size.height / 36.0f * 12.1f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Singapore.swift
//
//  UIImage+Singapore.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Singapore "wikipedia"
     
     Generates the flag of Singapore.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Singapore.
     */
    internal static func flagSingapore(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 238.0 / 255.0,
                               green: 37.0 / 255.0,
                               blue: 54.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: CGSize(width: size.width, height: size.height / 2.0)))

        colorWhite.setFill()
        var starRadius = size.height / 36.0 * 6.625
        UIBezierPath(arcCenter: CGPoint(x: size.width / 54.0 * 11.46, y: size.height / 36.0 * 9.0),
                     radius: starRadius,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorRed.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.width / 54.0 * 14.44, y: size.height / 36.0 * 9.0),
                     radius: starRadius,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorWhite.setFill()
        starRadius = size.height / 36.0 * 3.45 / 2.0
        // top
        UIBezierPath.star(at: CGPoint(x: size.width / 54.0 * 15.05, y: size.height / 36.0 * 5.2),
                          radius: starRadius).fill()
        // left middle
        UIBezierPath.star(at: CGPoint(x: size.width / 54.0 * 11.4, y: size.height / 36.0 * 7.85),
                          radius: starRadius).fill()
        // right middle
        UIBezierPath.star(at: CGPoint(x: size.width / 54.0 * 18.65, y: size.height / 36.0 * 7.85),
                          radius: starRadius).fill()
        // left bottom
        UIBezierPath.star(at: CGPoint(x: size.width / 54.0 * 12.8, y: size.height / 36.0 * 12.1),
                          radius: starRadius).fill()
        // right bottom
        UIBezierPath.star(at: CGPoint(x: size.width / 54.0 * 17.25, y: size.height / 36.0 * 12.1),
                          radius: starRadius).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
